### PR TITLE
Storage Component ECS

### PR DESCRIPTION
**Pull request by Fishfish458**
## About the PR 

This is a change to ECS storage component to remove a variety of bloat and change to use a bound interface.

At the moment this doesn't touch some other abstractions of sharedStorageComponent such as the entityStorageComponent, both are working in harmony right now and this change is getting big enough.

Opening this as a draft to start with since I'm trying to clear up anything that's left and wanted to see if anyone had any immediate comments for something missing.

As far as I've tested, all functionality of storage is still retained as long as the entities using the storage component have the storageUIkey added.

edit: actually looks like click pickup storages(trash bags) are not showing their animation or obeying their timer, working on that.

**Screenshots**

https://user-images.githubusercontent.com/47410468/163074040-fecddad1-2f59-4d9d-acf3-92a1f521dfc7.mp4

### CODE REVIEW

**Comment by Fishfish458** [comment]
Fixed the trash bag functionality

https://user-images.githubusercontent.com/47410468/163244395-e2532503-d7b2-4bd3-8234-7b8af69e81f8.mp4

https://user-images.githubusercontent.com/47410468/163244398-5ffcfdda-5ebb-42a3-9162-09024ffd6988.mp4

**Comment by Fishfish458** [comment]
Tested everything, should have exact same functionality as before, questionable areas is the clientinteractionsystem since we'd need to grab and check if the UI is open for that user on the frontend and the clientstoragecomponent.
Not sure if there's a better way to go about handling those since they're abstracting from sharedStorageComponent in which other components are still using.

**Comment by Fishfish458** [comment]
Just came to mind but one last issue is self destroying items are not handled, trying to figure that out now

**Comment by Fishfish458** [comment]
OK, actually have a working solution for the self deleting items now, thank you shadowcommander.

https://user-images.githubusercontent.com/47410468/165240857-ba1a4443-56a0-40f3-8503-0ced4ab0a249.mp4

I'm doing another run through but I believe this should be good for review now.

File: Content.Server/Storage/EntitySystems/StorageSystem.cs [hunk 1]
```diff
@@ -187,6 +191,234 @@ private void AddTransferVerbs(EntityUid uid, ServerStorageComponent component, G
...
+                CloseNestedInterfaces(uid, actor.PlayerSession, storageComp);
+
+            // If UI is closed for everyone
+            if (!_uiSystem.IsUiOpen(uid, args.UiKey))
+            {
+                storageComp.IsOpen = false;
+                UpdateStorageVisualization(uid, storageComp);
+
+                if (storageComp.StorageCloseSound is not null)
+                    SoundSystem.Play(Filter.Pvs(uid), storageComp.StorageCloseSound.GetSound(), uid, AudioParams.Default);
```
Comment (metalgearsloth) [inline Content.Server/Storage/EntitySystems/StorageSystem.cs]: Should pass in EntityManager to filters (they're kinda jank rn and should really be its own system).

File: Content.Server/Storage/EntitySystems/StorageSystem.cs [hunk 2]
```diff
@@ -236,57 +470,205 @@ private void AddTransferVerbs(EntityUid uid, ServerStorageComponent component, G
...
+
+            if (TryComp(insertEnt, out ServerStorageComponent? storage) &&
+                storage.StorageCapacityMax >= storageComp.StorageCapacityMax)
+                return false;
+
+            if (TryComp(insertEnt, out SharedItemComponent? itemComp) &&
+                itemComp.Size > storageComp.StorageCapacityMax - storageComp.StorageUsed)
+                return false;
+
+            if (storageComp.Whitelist != null && !storageComp.Whitelist.IsValid(insertEnt))
```
Comment (metalgearsloth) [inline Content.Server/Storage/EntitySystems/StorageSystem.cs]: Entitywhitelists also take in entitymanager too.

**Review by metalgearsloth — APPROVED** [review]
I already made the commented changes.

**Pull request merged**